### PR TITLE
HIVE-26976: Failure in direct SQL processing do not undo the changes made to database before fallback to JDO

### PR DESCRIPTION
**Pull request by VenuReddy2103**
### What changes were proposed in this pull request?
Setting savepoint in transaction before direct sql processing and rollback to savepoint before moving to JDO processing.

### Why are the changes needed?
Failure in direct SQL processing do not undo the changes made to database before falling back to JDO based processing. Thus results in stale/dangling entries in database forever.

For instance, during dropPartitions() direct SQL processing, after dropping rows from few tables(like PARTITIONS, PARTITION_PARAMS, PARTITION_KEY_VALS etc), it causes typecast exception in dropStorageDescriptors() and fallback to JDO processing. But datanucleus JDO processing cannot delete rows from remaining tables(i.e., from sds, serdes, sds_params, serde_params, sort_cols, bucketing_cols, skewed cols/values/location if any for the partitions) since the partition is already deleted in the same transaction during direct SQL processing.

This issue is applicable in the cases where database is modified(i.e., add partitions[https://github.com/apache/hive/pull/3905](https://github.com/apache/hive/pull/3905) , drop partitions) in direct sql processing and fails in between.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tested manually and also ran unittests

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 1]
```diff
@@ -668,6 +671,20 @@ public void rollbackTransaction() {
     }
   }
 
+  private void setTransactionSavePoint(String savePoint) {
+    if (savePoint != null) {
+      ExecutionContext ec = ((JDOPersistenceManager) pm).getExecutionContext();
+      ec.getStoreManager().getConnectionManager().getConnection(ec);
```
Comment (saihemanth-cloudera) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:677]: Should we try/catch null exception incase any of these becomes null?

Comment (VenuReddy2103) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:677]: I believe it is not required to check against null. These instances are created with the pm instance and cannot be null.

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:677]: Ack

**Review by saihemanth-cloudera — APPROVED** [review]
The changes look good to me. +1

**Pull request merged**